sirepo/template/ml.py

```diff
@@ -287,13 +287,83 @@ def stateful_compute_compute_column_info(data):
     return _compute_column_info(data.dataFile)
 
 
+def stateless_compute_get_remote_data(data):
+    return _get_remote_data(data.url, data.headers_only)
+
+
+def stateless_compute_remote_data_bytes_loaded(data):
+    return _remote_data_bytes_loaded(data.filename)
+
+
 def write_parameters(data, run_dir, is_parallel):
     pkio.write_text(
         run_dir.join(template_common.PARAMETERS_PYTHON_FILE),
         _generate_parameters_file(data),
     )
 
 
+def _build_model_py(v):
+    def _import_layers(v):
+        return "".join(", " + n for n in v.layerImplementationNames)
+
+    def _conv_args(layer):
+        if layer.layer not in ("Conv2D", "Transpose", "SeparableConv2D"):
+            return
+        return f"""{layer.dimensionality},
+    activation="{layer.activation}",
+    kernel_size=({layer.kernel}, {layer.kernel}),
+    strides={layer.strides},
+    padding="{layer.padding}"
+    """
+
+    def _pooling_args(layer):
+        return f'''pool_size=({layer.size}, {layer.size}),
+    strides={layer.strides},
+    padding="{layer.padding}"'''
+
+    args_map = PKDict(
```

Comment:
Should be defined before other functions.

Reply:
I'm not sure which one you mean. args_map?

Reply:
I think he means the args_map var should be first in the _build_model_py() method.

Alternatively, the _conv_args(), _pooling_args(), and args_map could be moved to a separate method:
```python
def _build_args_map():
    def _conv_args(layer):
        #...
    def _pooling_args(layer):
        #...
    return PKDict(
        #...
        AveragePooling2D=lambda layer: _pooling_args(layer),
        #...
    )
```

Reply:
> I think he means the args_map var should be first in the _build_model_py() method.

The args_map could be a constant, possibly declared at the global scope. That's kind of the point of having maps, otherwise a bunch of if's is probably more efficient and just as clear. 

`_conv_args` does not need to be defined in order to use it in a lambda definition. The name is bound at the time of the call, not at compile time of the lambda, just like for normal function definitions.

Defining it in a function just makes it more confusing.